File: src/SystemMonitorLogger/Platform/LinuxMetricsProvider.cs
using System.Globalization;
using SystemMonitorLogger.Models;

namespace SystemMonitorLogger.Platform;

public sealed class LinuxMetricsProvider : ISystemMetricsProvider
{
    private CpuTimes? _lastCpuTimes;
    private DiskIoSnapshot? _lastDiskIoSnapshot;

    public async Task<SystemSample> GetSystemSampleAsync(CancellationToken cancellationToken)
    {
        var now = DateTimeOffset.Now;
        var cpuPercent = await GetCpuPercentAsync(cancellationToken);
        var memory = await GetMemoryAsync(cancellationToken);
        var disk = GetPrimaryDisk();
        var diskIo = await GetDiskIoAsync(now, cancellationToken);

        return new SystemSample(
            now,
            cpuPercent,
            memory.UsedPercent,
            memory.UsedMb,
            memory.TotalMb,
            disk.UsedPercent,
            disk.FreeMb,
            disk.TotalMb,
            diskIo.ReadMbPerSecond,
            diskIo.WriteMbPerSecond);
    }

    private async Task<double> GetCpuPercentAsync(CancellationToken cancellationToken)
    {
        string? cpuLine = null;
        await foreach (var line in File.ReadLinesAsync("/proc/stat", cancellationToken))
        {
            cpuLine = line;
            break;
        }

        if (cpuLine is null)
        {
            return 0;
        }

        var firstLine = cpuLine.Split(' ', StringSplitOptions.RemoveEmptyEntries);

        if (firstLine.Length < 8 || firstLine[0] != "cpu")
        {
            return 0;
        }

        var user = Parse(firstLine[1]);
        var nice = Parse(firstLine[2]);
        var system = Parse(firstLine[3]);
        var idle = Parse(firstLine[4]);
        var iowait = Parse(firstLine[5]);
        var irq = Parse(firstLine[6]);
        var softirq = Parse(firstLine[7]);
        var steal = firstLine.Length > 8 ? Parse(firstLine[8]) : 0;

        var current = new CpuTimes(user + nice + system + irq + softirq + steal, idle + iowait);
        if (_lastCpuTimes is null)
        {
            _lastCpuTimes = current;
            return 0;
        }

        var previous = _lastCpuTimes.Value;
        _lastCpuTimes = current;

        var busyDelta = current.Busy - previous.Busy;
        var idleDelta = current.Idle - previous.Idle;
        var total = busyDelta + idleDelta;

        return total <= 0 ? 0 : Math.Clamp(busyDelta * 100d / total, 0, 100);
    }

    private static async Task<MemoryInfo> GetMemoryAsync(CancellationToken cancellationToken)
    {
        var values = new Dictionary<string, double>(StringComparer.OrdinalIgnoreCase);

        await foreach (var line in File.ReadLinesAsync("/proc/meminfo", cancellationToken))
        {
            var parts = line.Split(':', 2);
            if (parts.Length != 2)
            {
                continue;
            }

            var number = parts[1].Trim().Split(' ', StringSplitOptions.RemoveEmptyEntries).FirstOrDefault();
            if (number is not null && double.TryParse(number, NumberStyles.Float, CultureInfo.InvariantCulture, out var kb))
            {
                values[parts[0]] = kb;
            }
        }

        var totalMb = values.GetValueOrDefault("MemTotal") / 1024d;
        var availableMb = values.GetValueOrDefault("MemAvailable") / 1024d;
        var usedMb = Math.Max(0, totalMb - availableMb);
        var usedPercent = totalMb <= 0 ? 0 : usedMb * 100d / totalMb;

        return new MemoryInfo(totalMb, usedMb, usedPercent);
    }

    private static DiskInfo GetPrimaryDisk()
    {
        var drive = DriveInfo.GetDrives()
            .Where(d => d.IsReady && d.DriveType == DriveType.Fixed)
            .OrderByDescending(d => d.TotalSize)
            .FirstOrDefault()
            ?? DriveInfo.GetDrives().First(d => d.IsReady);

        var totalMb = drive.TotalSize / 1024d / 1024d;
        var freeMb = drive.AvailableFreeSpace / 1024d / 1024d;
        var usedPercent = totalMb <= 0 ? 0 : (totalMb - freeMb) * 100d / totalMb;

        return new DiskInfo(totalMb, freeMb, usedPercent);
    }

    private async Task<DiskIoInfo> GetDiskIoAsync(DateTimeOffset timestamp, CancellationToken cancellationToken)
    {
        var snapshot = await TryGetDiskIoSnapshotAsync(timestamp, cancellationToken);
        if (snapshot is null)
        {
            return new DiskIoInfo(0, 0);
        }

        if (_lastDiskIoSnapshot is null)
        {
            _lastDiskIoSnapshot = snapshot;
            return new DiskIoInfo(0, 0);
        }

        var previous = _lastDiskIoSnapshot.Value;
        _lastDiskIoSnapshot = snapshot;

        var elapsedSeconds = (snapshot.Value.Timestamp - previous.Timestamp).TotalSeconds;
        if (elapsedSeconds <= 0)
        {
            return new DiskIoInfo(0, 0);
        }

        var readDelta = snapshot.Value.ReadBytes >= previous.ReadBytes ? snapshot.Value.ReadBytes - previous.ReadBytes : 0;
        var writeDelta = snapshot.Value.WriteBytes >= previous.WriteBytes ? snapshot.Value.WriteBytes - previous.WriteBytes : 0;

        return new DiskIoInfo(
            readDelta / 1024d / 1024d / elapsedSeconds,
            writeDelta / 1024d / 1024d / elapsedSeconds);
    }

    private static async Task<DiskIoSnapshot?> TryGetDiskIoSnapshotAsync(DateTimeOffset timestamp, CancellationToken cancellationToken)
    {
        if (!File.Exists("/proc/diskstats"))
        {
            return null;
        }

        ulong readSectors = 0;
        ulong writtenSectors = 0;

        await foreach (var line in File.ReadLinesAsync("/proc/diskstats", cancellationToken))
        {
            var parts = line.Split(' ', StringSplitOptions.RemoveEmptyEntries);
            if (parts.Length < 14 || !IsPhysicalDisk(parts[2]))
            {
                continue;
            }

            readSectors += Parse(parts[5]);
            writtenSectors += Parse(parts[9]);
        }

        if (readSectors == 0 && writtenSectors == 0)
        {
            return null;
        }

        const ulong linuxSectorSize = 512;
        return new DiskIoSnapshot(timestamp, readSectors * linuxSectorSize, writtenSectors * linuxSectorSize);
    }

    private static bool IsPhysicalDisk(string name)
    {
        if (name.StartsWith("loop", StringComparison.OrdinalIgnoreCase)
            || name.StartsWith("ram", StringComparison.OrdinalIgnoreCase)
            || name.StartsWith("dm-", StringComparison.OrdinalIgnoreCase)
            || name.StartsWith("md", StringComparison.OrdinalIgnoreCase))
        {
            return false;
        }

        if (name.StartsWith("sd", StringComparison.OrdinalIgnoreCase) && name.Length == 3)
        {
            return true;
        }

        if (name.StartsWith("vd", StringComparison.OrdinalIgnoreCase) && name.Length == 3)
        {
            return true;
        }

        if (name.StartsWith("xvd", StringComparison.OrdinalIgnoreCase) && name.Length == 4)
        {
            return true;
        }

        if (name.StartsWith("nvme", StringComparison.OrdinalIgnoreCase) && name.EndsWith("n1", StringComparison.OrdinalIgnoreCase))
        {
            return true;
        }

        return false;
    }

    private static ulong Parse(string value) => ulong.Parse(value, CultureInfo.InvariantCulture);

    private readonly record struct CpuTimes(ulong Busy, ulong Idle);
    private readonly record struct MemoryInfo(double TotalMb, double UsedMb, double UsedPercent);
    private readonly record struct DiskInfo(double TotalMb, double FreeMb, double UsedPercent);
    private readonly record struct DiskIoInfo(double ReadMbPerSecond, double WriteMbPerSecond);
    private readonly record struct DiskIoSnapshot(DateTimeOffset Timestamp, ulong ReadBytes, ulong WriteBytes);
}

File: SystemMonitorLoggerCpp/src/platform/LinuxMetricsProvider.cpp
// Implementacao so faz sentido no Linux. Em outros alvos o arquivo compila vazio.
#if !defined(_WIN32)

#include "platform/LinuxMetricsProvider.hpp"
#include "utils/TimeHelper.hpp"

#include <sys/statvfs.h>

#include <algorithm>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>

namespace
{
    // Quebra uma linha em campos separados por espacos (ignora vazios).
    std::vector<std::string> SplitWhitespace(const std::string& line)
    {
        std::vector<std::string> parts;
        std::istringstream stream(line);
        std::string token;
        while (stream >> token)
        {
            parts.push_back(token);
        }
        return parts;
    }

    std::uint64_t ParseU64(const std::string& value)
    {
        try
        {
            return static_cast<std::uint64_t>(std::stoull(value));
        }
        catch (...)
        {
            return 0;
        }
    }

    // Filtra discos fisicos em /proc/diskstats (mesma regra do projeto C#).
    bool IsPhysicalDisk(const std::string& name)
    {
        auto startsWith = [&](const char* prefix) {
            return name.rfind(prefix, 0) == 0;
        };

        if (startsWith("loop") || startsWith("ram") || startsWith("dm-") || startsWith("md"))
        {
            return false;
        }

        if ((startsWith("sd") || startsWith("vd")) && name.size() == 3)
        {
            return true;
        }

        if (startsWith("xvd") && name.size() == 4)
        {
            return true;
        }

        if (startsWith("nvme") && name.size() >= 2 &&
            name.compare(name.size() - 2, 2, "n1") == 0)
        {
            return true;
        }

        return false;
    }
}

double LinuxMetricsProvider::GetCpuPercent()
{
    std::ifstream stat("/proc/stat");
    std::string line;
    if (!std::getline(stat, line))
    {
        return 0.0;
    }

    const auto parts = SplitWhitespace(line);
    if (parts.size() < 8 || parts[0] != "cpu")
    {
        return 0.0;
    }

    const std::uint64_t user = ParseU64(parts[1]);
    const std::uint64_t nice = ParseU64(parts[2]);
    const std::uint64_t system = ParseU64(parts[3]);
    const std::uint64_t idle = ParseU64(parts[4]);
    const std::uint64_t iowait = ParseU64(parts[5]);
    const std::uint64_t irq = ParseU64(parts[6]);
    const std::uint64_t softirq = ParseU64(parts[7]);
    const std::uint64_t steal = parts.size() > 8 ? ParseU64(parts[8]) : 0;

    const CpuTimes current{user + nice + system + irq + softirq + steal, idle + iowait};

    if (!m_lastCpu)
    {
        m_lastCpu = current;
        return 0.0;
    }

    const CpuTimes previous = *m_lastCpu;
    m_lastCpu = current;

    const std::uint64_t busyDelta = current.busy - previous.busy;
    const std::uint64_t idleDelta = current.idle - previous.idle;
    const std::uint64_t total = busyDelta + idleDelta;

    if (total == 0)
    {
        return 0.0;
    }

    const double percent = static_cast<double>(busyDelta) * 100.0 / static_cast<double>(total);
    return std::clamp(percent, 0.0, 100.0);
}

void LinuxMetricsProvider::GetMemory(double& totalMb, double& usedMb, double& usedPercent)
{
    double totalKb = 0.0;
    double availableKb = 0.0;

    std::ifstream meminfo("/proc/meminfo");
    std::string line;
    while (std::getline(meminfo, line))
    {
        const auto colon = line.find(':');
        if (colon == std::string::npos)
        {
            continue;
        }

        const std::string key = line.substr(0, colon);
        const auto numberParts = SplitWhitespace(line.substr(colon + 1));
        if (numberParts.empty())
        {
            continue;
        }

        const double kb = static_cast<double>(ParseU64(numberParts[0]));
        if (key == "MemTotal")
        {
            totalKb = kb;
        }
        else if (key == "MemAvailable")
        {
            availableKb = kb;
        }
    }

    totalMb = totalKb / 1024.0;
    const double availableMb = availableKb / 1024.0;
    usedMb = std::max(0.0, totalMb - availableMb);
    usedPercent = totalMb <= 0.0 ? 0.0 : usedMb * 100.0 / totalMb;
}

void LinuxMetricsProvider::GetDisk(double& totalMb, double& freeMb, double& usedPercent)
{
    struct statvfs vfs{};
    if (statvfs("/", &vfs) != 0)
    {
        totalMb = freeMb = usedPercent = 0.0;
        return;
    }

    const double blockSize = static_cast<double>(vfs.f_frsize);
    totalMb = static_cast<double>(vfs.f_blocks) * blockSize / 1024.0 / 1024.0;
    freeMb = static_cast<double>(vfs.f_bavail) * blockSize / 1024.0 / 1024.0;
    usedPercent = totalMb <= 0.0 ? 0.0 : (totalMb - freeMb) * 100.0 / totalMb;
}

void LinuxMetricsProvider::GetDiskIo(double& readMbPerSecond, double& writeMbPerSecond, double& activePercent)
{
    readMbPerSecond = 0.0;
    writeMbPerSecond = 0.0;
    activePercent = 0.0;

    std::ifstream diskstats("/proc/diskstats");
    if (!diskstats.is_open())
    {
        return;
    }

    std::uint64_t readSectors = 0;
    std::uint64_t writtenSectors = 0;
    // io_ticks (ms ocupado) por disco, para o %util desta amostra.
    std::map<std::string, std::uint64_t> ioTicks;
    std::string line;
    while (std::getline(diskstats, line))
    {
        const auto parts = SplitWhitespace(line);
        if (parts.size() < 14 || !IsPhysicalDisk(parts[2]))
        {
            continue;
        }

        readSectors += ParseU64(parts[5]);
        writtenSectors += ParseU64(parts[9]);
        // Campo 13 (indice 12): "ms doing I/O" = io_ticks, base do %util.
        ioTicks[parts[2]] = ParseU64(parts[12]);
    }

    constexpr std::uint64_t kSectorSize = 512;
    const DiskIoSnapshot current{
        std::chrono::steady_clock::now(),
        readSectors * kSectorSize,
        writtenSectors * kSectorSize};

    const bool hadPrevious = m_lastDiskIo.has_value();
    const DiskIoSnapshot previous = hadPrevious ? *m_lastDiskIo : current;
    m_lastDiskIo = current;

    // %util: maior percentual entre os discos fisicos (o disco mais ocupado).
    // util = delta(io_ticks_ms) / tempo_decorrido_ms * 100.
    const double elapsedSeconds =
        std::chrono::duration<double>(current.timestamp - previous.timestamp).count();
    if (hadPrevious && elapsedSeconds > 0.0)
    {
        const double elapsedMs = elapsedSeconds * 1000.0;
        double maxUtil = 0.0;
        for (const auto& [name, ticks] : ioTicks)
        {
            const auto it = m_lastIoTicks.find(name);
            if (it == m_lastIoTicks.end())
            {
                continue;
            }
            const std::uint64_t deltaMs = ticks >= it->second ? ticks - it->second : 0;
            maxUtil = std::max(maxUtil, static_cast<double>(deltaMs) / elapsedMs * 100.0);
        }
        activePercent = std::clamp(maxUtil, 0.0, 100.0);
    }
    m_lastIoTicks = std::move(ioTicks);

    if (!hadPrevious || elapsedSeconds <= 0.0)
    {
        return;
    }

    const std::uint64_t readDelta =
        current.readBytes >= previous.readBytes ? current.readBytes - previous.readBytes : 0;
    const std::uint64_t writeDelta =
        current.writeBytes >= previous.writeBytes ? current.writeBytes - previous.writeBytes : 0;

    readMbPerSecond = static_cast<double>(readDelta) / 1024.0 / 1024.0 / elapsedSeconds;
    writeMbPerSecond = static_cast<double>(writeDelta) / 1024.0 / 1024.0 / elapsedSeconds;
}

SystemSample LinuxMetricsProvider::GetSystemSample()
{
    SystemSample sample;
    sample.timestamp = TimeHelper::NowString();
    sample.cpuPercent = GetCpuPercent();
    GetMemory(sample.memoryTotalMb, sample.memoryUsedMb, sample.memoryUsedPercent);
    GetDisk(sample.diskTotalMb, sample.diskFreeMb, sample.diskUsagePercent);
    GetDiskIo(sample.diskReadMbPerSecond, sample.diskWriteMbPerSecond, sample.diskActivePercent);
    return sample;
}

#endif // !_WIN32
